### ISSUE TITLE
fastspeech inference duration_outputs

### ISSUE THREAD

**Issue by superhg2012** [dathudeptrai/TensorflowTTS#30]
when run fastspeech in inference mode, why  use exp op after duration_output from duration predictor?

`duration_outputs = tf.math.exp(duration_outputs) - 1.0`

**Comment by dathudeptrai** [comment]
@superhg2012 see https://github.com/dathudeptrai/TensorflowTTS/blob/master/examples/fastspeech/train_fastspeech.py#L126-L127. I use log loss for duration.

**Comment by superhg2012** [comment]
thanks @dathudeptrai  Also,  your params and activation are different with paper configs. Does mish activation and relu6 activation helps improve synthesized audio quality? Did you try params in fastspeech paper?

**Comment by dathudeptrai** [comment]
@superhg2012 it improve a loss.

**Comment by superhg2012** [comment]
> @superhg2012 it improve a loss.

a lot?

**Comment by dathudeptrai** [comment]
from 0.25 to 0.23 mae loss :))

**Comment by superhg2012** [comment]
I have two questions
1 dropout in postnet set to 0.1 rather than 0.5?
2 why use mask in each model and layer?

**Comment by dathudeptrai** [comment]
> dropout in postnet set to 0.1 rather than 0.5?

I don't see any different between mel loss before and after postnet. Maybe ignore postnet is sill ok. For dropout i set it = 0.1 rather than 0.5 because i am not aware of it :)). Maybe 0.5 is better ^^.

> why use mask in each model and layer?

Just for safe, that is my habit. Mask every where to make sure everything work correctly, i don't really want there is a bug related with "padding".

**Comment by superhg2012** [comment]
In multi-head self-attention part, why there is a dense layer before dropout and Layernorm? what's it for?

**Comment by superhg2012** [comment]
dk in self-attention softmax  is 64 in most fastspeech implementations, does it affect too much?

**Comment by dathudeptrai** [comment]
@superhg2012 that is a thing i missed. But it's not important, in my implementation, 4 layers is enough, if use df=64, use should use 6 layers to get a comparable model with my models.

**Comment by superhg2012** [comment]
I am training fastspeech with post-cbhg for linear_spectrogram predicting, and also tried 6 and 4 FFT layers on a 6 hours dataset, on training dataset , audio is perfect ,but bad quality on test texts.  Also I am using noam learning rate scheduling. my loss is 0.6

**Comment by dathudeptrai** [comment]
Can u share ur learning curves, loss 0.6 seem too high, my loss around 0.3

**Comment by superhg2012** [comment]
![image](https://user-images.githubusercontent.com/8817309/83863746-5ac35c00-a756-11ea-8d5b-3d6f7baf2a5b.png)

**Comment by dathudeptrai** [comment]
@superhg2012  are u using my implementation ?

**Comment by superhg2012** [comment]
not yet, I wirte it with tensorflow1.0. I have post-cbhg for linear-spectrogram.  Does normalization for mel-spectrogram work?

**Comment by dathudeptrai** [comment]
yes it work :D

**Comment by superhg2012** [comment]
where is your alignments  from?  Tacotron2 with r=1? I found that alignment data affects fastspeech training too much

**Comment by dathudeptrai** [comment]
@superhg2012 yes, i use my tacotron-2 at 40K to extract durations. I will re-train fs with tacotron-2 at 120k steps with window masking. Yeah, duration affect fs too much.

**Comment by superhg2012** [comment]
yes, I see that in your tacotron2, teacher forcing ratio is constant 1.0 during training and during testing(extract durarion)  use  prev_predicted_outputs all the time.

**Comment by dathudeptrai** [comment]
> yes, I see that in your tacotron2, teacher forcing ratio is constant 1.0 during training and during testing(extract durarion) use prev_predicted_outputs all the time

yes :)), i implement teacher forcing ratio for people who want to try :)). But training without is is enough to get good model :D.

**Comment by superhg2012** [comment]
I used teacher foring scheduled for training and extract alignment with gta

**Comment by dathudeptrai** [comment]
@superhg2012 i think the best workflow is: training tacotron without scheduled teacher forcing ->>> extract alignment with gta and window masking to force monotonic alignment.

**Comment by superhg2012** [comment]
@dathudeptrai  thanks , I will try it
